Use SST and SSS consistent with MARBL tracers

mnlevy1981 opened the pull request:
When computing surface fluxes and interior tendencies for MARBL tracers, we want to use T & S from before they are modified by `tracer_vertdiff()`. We do this by adding `prediabatic_T` and `prediabatic_S` to the diabatic control structure, and then passing these values instead of `tv%T` and `tv%S` to MARBL.

mnlevy1981 committed:
Use prediabatic SST and SSS for MARBL surface flux

The existing MARBL driver uses T & S from the end of the diabatic routine as
inputs into surface_flux_compute() and interior_tendency_compute(). For
surface_flux_compute(), there is a discrepancy because we have not yet called
tracer_vertdiff for the MARBL tracers but have for T & S. To enforce
consistency among all the tracers, we want to use the pre-vertdiff T&S
quantities.

There is a little bit of infrastructure change -- I added 3D fields for T & S
at the beginning of the diabatic subroutine to the diabatic control structure
(we will eventually want to call interior_tendency_compute() before
tracer_vertdiff as well), and then had to add a routine to
MOM_tracer_flow_control.F90 to extract use_MARBL_tracers from the tracer flow
control structure. prediabatic_T and prediabatic_S are now optional arguments
to call_tracer_column_fns(), but that function will abort if USE_MARBL_TRACERS
is true and the arguments are not present because they are required in
MARBL_tracers_column_physics()

mnlevy1981 committed:
Clean up comments

Fix failing doxygen / style check texts. Also add a long comment to
diabatic_ALE explaining why we use prediabatic_T and prediabatic_S (maybe the
comment belongs in diabatic_init?)

mnlevy1981 commented:
As a first pass, I only used `prediabatic_T` and `prediabatic_S` for the surface flux computation; I will mark this as "ready for review" when we are also using these arrays to compute the interior tendency (which will require reordering some calls in `MARBL_tracers_column_physics()`)

mnlevy1981 committed:
Use prediabatic T & S for MARBL interior tendency

Reordered MARBL_tracers_column_physics so that surface_flux_compute() and
interior_tendency_compute() are both called before tracer_vertdiff. Now use
prediabatic T & S for interior tendency computation, and also use h_old as dz.

Note that this greatly increases the number of warnings from the co2calc
routine, and we are not sure why.

mnlevy1981 commented:
@gustavo-marques is using these changes in his latest runs where we will look at the effect of the two different vertical grids on BGC variables; assuming those runs don't raise any red flags, this PR is ready to go (it will also require https://github.com/marbl-ecosys/MARBL/pull/480)

klindsay28 commented on src/tracer/MOM_tracer_flow_control.F90:
```diff
@@ -527,13 +543,17 @@ subroutine call_tracer_column_fns(h_old, h_new, ea, eb, fluxes, mld, dt, G, GV,
                                         nonLocalTrans=nonLocalTrans, &
                                         evap_CFL_limit=evap_CFL_limit, &
                                         minimum_forcing_depth=minimum_forcing_depth)
+    end if
```
MOM6 style guide says to use `endif` instead of `end if`.
There are 3 occurrences of `end if` in this PR.

klindsay28 commented on src/tracer/MOM_tracer_flow_control.F90:
```diff
@@ -415,6 +415,17 @@ subroutine get_chl_from_model(Chl_array, G, GV, CS)
+  ! All output arguments are optional
+  logical,                    optional, intent(out) :: use_MARBL_tracers  !< If true, MARBL tracers are active
+
+  ! Constants within diabatic_CS
```
This comment appears to have been cut and pasted from `extract_diabatic_member`.
It needs to be modified to match this context.

klindsay28 commented on src/tracer/MOM_tracer_flow_control.F90:
```diff
@@ -201,7 +201,7 @@ subroutine call_tracer_register(G, GV, US, param_file, CS, tr_Reg, restart_CS)
                  "If true, use the ideal_age_example tracer package.", &
                  default=.false.)
-  call get_param(param_file, mdl, "USE_MARBL_TRACERS", CS%use_marbl_tracers, &
+  call get_param(param_file, mdl, "USE_MARBL_TRACERS", CS%use_MARBL_tracers, &
```
This edit is making the case of `use_MARBL_tracers` consistent in the file `MOM_tracer_flow_control.F90`.
This is good.

There are other occurrences of `use_marbl_tracers` across MOM6.
Should we make the case of this consistent across all MOM6 source code?

mnlevy1981 committed:
Clean up following code review

1. removed spaces from "end if" to match style guide recommendations
2. cleaned up comment in extract_tracer_flow_member (refer to
   tracer_flow_control_CS rather than diabatic_CS)
3. Consistent capitalization of MARBL in use_MARBL_tracers (and use_MARBL)
   throughout the codebase

klindsay28 reviewed (approved):
LGTM

alperaltuntas commented:
@mnlevy1981  Has this PR been tested, or is it ready for being tested?

mnlevy1981 commented:
@alperaltuntas it is ready for testing (I'm surprised I don't have a comment mentioning test results -- I'm pretty sure I ran at least `aux_mom_MARBL` but I can't find any record of it)

note that this is answer-changing for cases with `USE_MARBL_TRACERS=TRUE`

gustavo-marques reviewed (approved):
LGTM!

alperaltuntas commented:
@mnlevy1981 For both MARBL test cases in aux_mom, I got ~35% performance hit. Is this expected?

```
    FAIL SMS.TL319_t232.G1850MARBL_JRA.derecho_intel TPUTCOMP Error: TPUTCOMP: Throughput changed by 35.14%: baseline=4.963 sypd, tolerance=25%, current=3.219 sypd
```

mnlevy1981 commented:
> @mnlevy1981 For both MARBL test cases in aux_mom, I got ~35% performance hit. Is this expected?
> 
> ```
>     FAIL SMS.TL319_t232.G1850MARBL_JRA.derecho_intel TPUTCOMP Error: TPUTCOMP: Throughput changed by 35.14%: baseline=4.963 sypd, tolerance=25%, current=3.219 sypd
> ```

Can you try updating MARBL to the head of the `development` branch? These changes greatly increased the number of MARBL warnings that were being written to `cesm.log`, and [marbl-ecosys/MARBL#480](https://github.com/marbl-ecosys/MARBL/pull/480) brought that warning count back down to a reasonable number. I'm a little surprised it was a 35% hit, but I want to make a new MARBL tag for `MOM-interface` with these changes anyway.

alperaltuntas commented:
> > @mnlevy1981 For both MARBL test cases in aux_mom, I got ~35% performance hit. Is this expected?
> > ```
> >     FAIL SMS.TL319_t232.G1850MARBL_JRA.derecho_intel TPUTCOMP Error: TPUTCOMP: Throughput changed by 35.14%: baseline=4.963 sypd, tolerance=25%, current=3.219 sypd
> > ```
> 
> Can you try updating MARBL to the head of the `development` branch? These changes greatly increased the number of MARBL warnings that were being written to `cesm.log`, and [marbl-ecosys/MARBL#480](https://github.com/marbl-ecosys/MARBL/pull/480) brought that warning count back down to a reasonable number. I'm a little surprised it was a 35% hit, but I want to make a new MARBL tag for `MOM-interface` with these changes anyway.

Updating MARBL to the head of development brought throughput to 4.77.
